### PR TITLE
EVG-8024 Support passing file to CLI to create hosts

### PR DESCRIPTION
**Pull request by malikchaya2**
[EVG-8024](https://jira.mongodb.org/browse/EVG-8024)

This will allow creating a spawn hosts with the CLI using a document. Instead of creating a host using a command like this: 
`evergreen host create -d rhel76-small -t a=b -k key -s testUserData.sh`
one can now create a host like this: 
`evergreen host create -f host-test.yml` 
with the following in host-test.yml: 
```distro: rhel76-small 
region: us-east-1 
key: key
tag: a=b
userdata_file: testUserData.sh
``` 

There are some restrictions: 
1. This only allows for one host to be created at a time. 
2. The ultimate goal is for the same file to be used to create a host from the CLI and using host.create in the yaml file (implemented[ here](https://jira.mongodb.org/browse/EVG-7915)). However, [CreateHost](https://github.com/evergreen-ci/evergreen/blob/master/apimodels/agent_models.go#L104) is a lot more robust than [HostRequestOptions](https://github.com/evergreen-ci/evergreen/blob/master/rest/model/host.go#L32 ). The overlap between the two are: 
```
distro
userdata_file
region
```

### CODE REVIEW

File: operations/host_spawn.go [hunk 1]
```diff
@@ -68,6 +69,10 @@ func hostCreate() cli.Command {
 				Name:  noExpireFlagName,
 				Usage: "make host never expire",
 			},
+			cli.StringFlag{
+				Name:  joinFlagNames(fileFlagName, "f"),
+				Usage: "name of a file containing the span host params",
```
Comment (ablack12) [inline operations/host_spawn.go:74]: nit: spawn

File: operations/host_spawn.go [hunk 2]
```diff
@@ -89,29 +95,42 @@ func hostCreate() cli.Command {
...
 			defer client.Close()
 
-			var script string
+			spawnRequest := &restModel.HostRequestOptions{}
+			tags := []host.Tag 
+			if file != "" {
+				err = utility.ReadYAMLFile(file, &spawnRequest)
+				if err != nil {
+					return errors.Wrapf(err, "problem reading from file '%s'", file)
+				}
```
Comment (ablack12) [inline operations/host_spawn.go:104]: I realize it's messy that host.create uses one interface and spawn hosts use another. Since the point of this file is to allow `host.create` scripts to be passed into the CLI, I think it's okay to restrict this to only the subset of host.create parameters that we want to support, and since tags isn't a host.create parameter we don't need to handle it in the file.

File: operations/host_spawn.go [hunk 3]
```diff
@@ -89,29 +95,42 @@ func hostCreate() cli.Command {
 			client := conf.setupRestCommunicator(ctx)
 			defer client.Close()
 
-			var script string
+			spawnRequest := &restModel.HostRequestOptions{}
+			tags := []host.Tag
```
Comment (ablack12) [inline operations/host_spawn.go:99]: No reason to initialize this outside of the if/else since each handles it separately.

File: operations/host_spawn.go (hunk 2 above)
Comment (malikchaya2) [reply operations/host_spawn.go:104]: That is a good point. However, since it's already working, I see no reason to remove the functionality since someone might have a different use case and find it useful. Additionally, not supporting it will end up requiring more documentation and potentially open more questions from users down the line. Thoughts?

File: operations/host_spawn.go (hunk 3 above)
Comment (malikchaya2) [reply operations/host_spawn.go:99]: thanks!

File: operations/host_spawn.go (hunk 1 above)
Comment (malikchaya2) [reply operations/host_spawn.go:74]: thank you!

File: operations/host_spawn.go (hunk 3 above)
Comment (malikchaya2) [reply operations/host_spawn.go:99]: Ah I remember now why I did that. The linter made me define it separately because otherwise err shadows the above definition. Since I need to define it separately anyhow, and it will end up being used one way or another (so there's no extra overhead), it feels cleaner to define it once instead of defining it twice.

File: apimodels/agent_models.go [hunk 4]
```diff
@@ -120,7 +120,7 @@ type CreateHost struct {
 	SecurityGroups  []string    `mapstructure:"security_group_ids" json:"security_group_ids" yaml:"security_group_ids" plugin:"expand"`
 	Spot            bool        `mapstructure:"spot" json:"spot" yaml:"spot"`
 	Subnet          string      `mapstructure:"subnet_id" json:"subnet_id" yaml:"subnet_id" plugin:"expand"`
-	UserdataFile    string      `mapstructure:"userdata_file" json:"-" yaml:"-" plugin:"expand"`
+	UserdataFile    string      `mapstructure:"userdata_file" json:"-" yaml:"script" plugin:"expand"`
```
Comment (ablack12) [inline apimodels/agent_models.go:123]: Why was this changed to script?

File: operations/host_spawn.go (hunk 2 above)
Comment (ablack12) [reply operations/host_spawn.go:104]: Okay sounds good, as long as the documentation is clear!

File: apimodels/agent_models.go (hunk 4 above)
Comment (malikchaya2) [reply apimodels/agent_models.go:123]: So that it is consistent between host.create and the CLI.

Comment (ablack12) [reply apimodels/agent_models.go:123]: I'm wary of this change because it could break existing yaml files that use create.host and expect `userdata_file`. This sort of goes along with what I was saying about tags: the data that's included in the file should work for how host.create is expected now; we don't want to make too many allowances/changes for how the CLI works now.

Comment (malikchaya2) [reply apimodels/agent_models.go:123]: Is this still a concern considering that I added the yaml tags (in EVG-7915)? Therefore there shouldn't be an existing use case for someone using yaml with "-" instead of "script". If I don't make this change, then the people who wan't to use the same file to spin up a host from host.create and the CLI won't be able to define a userData script.

Comment (ablack12) [reply apimodels/agent_models.go:123]: I guess to "request changes" I need to comment here again rip sorry but yeah, better to be consistent with how host.create works now than with flag names

Comment (ybrill) [reply apimodels/agent_models.go:123]: would it work to case on whether the user has provided parameters vs. a file. If they provide parameters we do what we've been doing. If they pass a file we read it into a `CreateHost` and call [CreateHost](https://github.com/evergreen-ci/evergreen/blob/f66132cf5634d302c6ee8afc0666f65258d9ff15/rest/client/methods.go#L700) instead of [CreateSpawnHost](https://github.com/evergreen-ci/evergreen/blob/f66132cf5634d302c6ee8afc0666f65258d9ff15/rest/client/rest.go#L24). This way any file that works for host.create would do the same thing from the CLI as it does for host.create.

Comment (malikchaya2) [reply apimodels/agent_models.go:123]: followed up on slack

Comment (malikchaya2) [reply apimodels/agent_models.go:123]: There already is a separate CLI command to spawn from a file, but the params still get read into the same struct. I can work on implementing "If they pass a file we read it into a CreateHost and call CreateHost instead of CreateSpawnHost" but that would be adding additional work to something that's already working. It's probably doable, but the question is if it's worth it.

Comment (ybrill) [reply apimodels/agent_models.go:123]: up to you 😃

Comment (malikchaya2) [reply apimodels/agent_models.go:123]: I don't think it's worth it right now, but it may come up again and be worth it if they end up needing more parameters.

Comment (malikchaya2) [reply apimodels/agent_models.go]: I realize now, I think I was wrong when I said "There already is a separate CLI command to spawn from a file". I added a flag, not a separate CLI command." I'll follow up on slack.

File: rest/model/host.go [hunk 5]
```diff
@@ -30,19 +30,20 @@ type APIHost struct {
 
 // HostPostRequest is a struct that holds the format of a POST request to /hosts
 type HostRequestOptions struct {
```
Comment (ablack12) [inline rest/model/host.go:32]: Can you drop a comment to explain that the yaml tags are pulled from CreateHost so we understand why they are different in the future?

File: operations/host_spawn.go [hunk 6]
```diff
@@ -89,29 +95,42 @@ func hostCreate() cli.Command {
 			client := conf.setupRestCommunicator(ctx)
 			defer client.Close()
 
-			var script string
+			spawnRequest := &restModel.HostRequestOptions{}
+			var tags []host.Tag
+			if file != "" {
+				err = utility.ReadYAMLFile(file, &spawnRequest)
+				if err != nil {
```
Comment (ablack12) [inline operations/host_spawn.go:102]: nit: `if err =  utility.ReadYAMLFile(file, &spawnRequest); err != nil { `

**Comment by ablack12** [comment]
Also don't forget to remove WIP from the commit message before merging! (we also don't usually put a colon between the ticket and the message, i.e. "EVG-8024 support passing file to ... ")
